swarm/pss/forwarding_test.go

```diff
@@ -14,95 +14,122 @@ import (
 		// send msg directly to the known peers (recipient address == peer address)
 		testForwardMsg(100+i, t, ps, peerAddresses[i][:], peerAddresses, []int{i})
 	}
 
 	for i := 0; i < firstNearest; i++ {
-		// send random messages with different proximity orders
-		po := i / 2
-		dst := pot.RandomAddressAt(base, po)
-		testForwardMsg(200+i, t, ps, dst[:], peerAddresses, []int{po * 2, po*2 + 1})
+		// send random messages with proximity orders, corresponding to PO of each bin
+		a = pot.RandomAddressAt(base, i)
+		testForwardMsg(200+i, t, ps, a[:], peerAddresses, []int{i})
 	}
 
 	for i := firstNearest; i < len(peerAddresses); i++ {
 		// recipient address falls into the nearest neighbours' bin
-		dst := pot.RandomAddressAt(base, i)
-		testForwardMsg(300+i, t, ps, dst[:], peerAddresses, nearestNeighbours)
+		a = pot.RandomAddressAt(base, i)
+		testForwardMsg(300+i, t, ps, a[:], peerAddresses, nearestNeighbours)
 	}
 
-	// send msg with proximity order higher than the last nearest neighbour
-	dst = pot.RandomAddressAt(base, 29)
-	testForwardMsg(400, t, ps, dst[:], peerAddresses, nearestNeighbours)
+	// send msg with proximity order much deeper than the deepest nearest neighbour
+	a = pot.RandomAddressAt(base, 77)
+	testForwardMsg(400, t, ps, a[:], peerAddresses, nearestNeighbours)
 
 	// test with partial addresses
 	const part = 12
 
 	for i := 0; i < firstNearest; i++ {
 		// send messages with partial address falling into different proximity orders
-		po := i / 2
-		if po%8 != 0 {
-			testForwardMsg(500+i, t, ps, peerAddresses[i][:po], peerAddresses, []int{po * 2, po*2 + 1})
+		if i%8 != 0 {
+			testForwardMsg(500+i, t, ps, peerAddresses[i][:i], peerAddresses, []int{i})
 		}
-		testForwardMsg(550+i, t, ps, peerAddresses[i][:part], peerAddresses, []int{po * 2, po*2 + 1})
+		testForwardMsg(550+i, t, ps, peerAddresses[i][:part], peerAddresses, []int{i})
 	}
 
 	for i := firstNearest; i < len(peerAddresses); i++ {
 		// partial address falls into the nearest neighbours' bin
 		testForwardMsg(600+i, t, ps, peerAddresses[i][:part], peerAddresses, nearestNeighbours)
 	}
 
-	// partial address with proximity order higher than the last nearest neighbour
-	dst = pot.RandomAddressAt(base, part)
-	testForwardMsg(700, t, ps, dst[:part], peerAddresses, nearestNeighbours)
+	// partial address with proximity order deeper than any of the nearest neighbour
+	a = pot.RandomAddressAt(base, part)
+	testForwardMsg(700, t, ps, a[:part], peerAddresses, nearestNeighbours)
 
 	// special cases where partial address matches a large group of peers
-	all := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20}
+	all := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
 	testForwardMsg(800, t, ps, []byte{}, peerAddresses, all)
-	testForwardMsg(900, t, ps, peerAddresses[19][:1], peerAddresses, all[16:])
+
+	// luminous radius of one byte (8 bits)
+	testForwardMsg(900, t, ps, baseAddrBytes[:1], peerAddresses, all[8:])
```

Comment:
so we are missing a lot of cases: 

- test which shows that within the bin, the forward function selects the closest peer to the address first
- test which shows that if luminosity is < depth but not 0 then it sends to all peers deeper than luminosity
- test which shows what happens if send returns false (send error)

Therefore I really suggest using a pattern. 

```golang
type testCase struct{
   name string
    recipient []byte
    errors: []int
    expected: []int
    peers:  [][]byte
}

testCases := []{
{
  expected: all,
  peers: peers,
},
...
{
  recipient: RandomAddrAt(pivot, 2),
  errors: []bool{false, true},
  expected: []int{3},
  peers: peers, 
}
...
for _, tc := range testCases {
t.Run(tc.Name, func(t *testing.T) {
    if err := testForward(tc); err != nil {
       t.Fatal(err)
    }
}
```

```golang
func testForwardMsg(tc testCase) error {
   testSendFunc := func(...
     // use tc fields
    }
   // set sendFunc, defer reset
   // call forward on pss
}
```